### PR TITLE
Precompute global iris stats

### PR DESCRIPTION
**Pull request by diegocepedaw**
Add the option to precompute global iris stats and retrieve them from the db instead of calculating them for every request.

### CODE REVIEW

File: src/iris/app_stats.py [hunk 1]
```diff
@@ -133,3 +133,62 @@ def calculate_app_stats(app, connection, cursor, fields_filter=None):
         if count_stat not in stats:
             stats[count_stat] = 0
     return stats
+
+
+def calculate_gobal_stats(connection, cursor, fields_filter=None):
```
Comment (dwang159) [inline src/iris/app_stats.py]: global

File: db/schema_0.sql [hunk 2]
```diff
@@ -749,6 +749,22 @@ CREATE TABLE `application_stats` (
     ON DELETE CASCADE ON UPDATE CASCADE
 ) ENGINE=InnoDB DEFAULT CHARSET=latin1;
 
+DROP TABLE IF EXISTS `global_stats`;
```
Comment (dwang159) [inline db/schema_0.sql]: model on `application_stats` instead, so we can add global stats without needing a schema change.

File: src/iris/api.py [hunk 3]
```diff
@@ -4002,67 +4003,43 @@ def on_get(self, req, resp):
...
+                WHERE `timestamp` = (SELECT MAX(`timestamp`) from `global_stats` gs2 WHERE gs2.`timestamp` = gs1.`timestamp`)
+            '''
+            cursor.execute(query)
+
+            if cursor.rowcount == 0:
+                logger.error('Error retrieving global stats from db')
+                raise HTTPInternalServerError('Error retrieving global stats from db')
+
+            # build dictionary of stat names and values
+            desc = cursor.description
```
Comment (jrgp) [inline src/iris/api.py]: If you do `cursor = connection.cursor(db.dict_cursor)`, the rows cursor.fetchall() returns will be a dict mapping column names to values, which alleviates having to construct that dict manually here.

But then, that likely breaks the rest of the uses of this cursor as those all expect to not be using a dict cursor.

As you already know the fields which will be returned, as you listed them explicitly in your query, you could avoid relying on this metadata and construct the dict manually. eg:

```
columns = ('median_seconds_to_claim_last_month', 'total_incidents_today', 'total_active_users', 'total_plans', 'total_messages_sent', 'total_applications', 'pct_incidents_claimed_last_month', 'total_incidents', 'total_messages_sent_today', 'timestamp')
row = cursor.fetchone()
stats = {column: row[i] for column, i in enumerate(columns)}
```

File: src/iris/api.py [hunk 4]
```diff
@@ -4002,67 +4003,43 @@ def on_get(self, req, resp):
...
+        if self.real_time:
+            stats = app_stats.calculate_gobal_stats(connection, cursor, fields_filter=fields_filter)
+
+        else:
+            # select the most current global stats entry
+            query = '''
+                SELECT `median_seconds_to_claim_last_month`, `total_incidents_today`, `total_active_users`, `total_plans`, `total_messages_sent`,
+                `total_applications`, `pct_incidents_claimed_last_month`, `total_incidents`, `total_messages_sent_today`, `timestamp`
+                FROM `global_stats` gs1
+                WHERE `timestamp` = (SELECT MAX(`timestamp`) from `global_stats` gs2 WHERE gs2.`timestamp` = gs1.`timestamp`)
```
Comment (jrgp) [inline src/iris/api.py]: you can avoid this subquery and where clause by sorting on the timestamp descending. eg  `ORDER BY `timestamp desc` DESC LIMIT 1`.

**Pull request merged**